### ISSUE TITLE
[Question] Is there any way to freeze layers in a DNN model?

### ISSUE THREAD

**Issue by xiaolin-cheng** [davisking/dlib#1559]
Hi Davis,

I am working on incremental learning for face recognition. Based on the error samples predicted by a trained model, I would like to retrain the model, but not the whole thing, just probably last a few layers. Is there any way to freeze the parameters of a particular layer? 

Thank you!

**Comment by davisking** [comment]
Each layer allows you to set its learning rate multiplier.  You can set them to 0 if you want.

**Comment by davisking** [comment]
However, it's worth pointing out that if you are just freezing the last layer you really ought to just extract all the features to disk once and then use something like a SVM solver.  I say this because DNN solvers are awful, in the grand scheme of things.  SVMs use solvers that are many orders of magnitude faster and more accurate.

**Comment by xiaolin-cheng** [comment]
@davisking Thanks a lot! 

By "solver", did you refer to "classifier" to use features to make prediction? DNN solvers mean fully-connected layers after convolutional layers? A bit confused on your term of "solver".

I found the API learning_rate_multiplier in layers.h. Could you show me a quick example how to use it in the face recognition training model file: **dnn_metric_learning_on_images_ex.cpp**? 

Thank you very much! Really appreciate your great help.

**Comment by davisking** [comment]
The entire API is documented, you should start with the introductory DNN example programs and then read the reference API docs if you don't find what you need.

But to summarize, just call the member function on the layer to set the learning rate.

**Comment by xiaolin-cheng** [comment]
Hi Davis, @davisking 

I read the API docs and tried to figure out how to set lr multiplier:
**net.subnet().layer_details().set_learning_rate_multiplier(0.1);**

But net.subnet().subnet() recursively won't scale for a large network. Is there any way to iterate all computation layers of a network?

Thank you!

**Comment by davisking** [comment]
Read the introduction: http://dlib.net/dnn_introduction2_ex.cpp.html

Then consult the API reference where everything is documented. 

You can also look at routines like visit_layer_parameters to see how to make a loop over layers if you don't know how to make C++ compile time loops.

**Comment by xiaolin-cheng** [comment]
Hi Davis, @davisking 

Thank you very much for your helpful pointers. I figured it out how to access a layer.

I have the following code to look into the first conv layer of a ResNet18 network:
```
        cout << "\n******** Layer 79 ********\n";
        layer<79>(net).layer_details().set_learning_rate_multiplier(0.1);
        cout << layer<79>(net).layer_details() << "\n"; //con
        tensor& con_params = layer<79>(net).layer_details().get_layer_params();
        cout << "con size: " << con_params.size() << "\n";
        cout << "n: " << con_params.num_samples();
        cout << ", k: " << con_params.k();
        cout << ", nc: " << con_params.nc();
        cout << ", nr: " << con_params.nr();
        con_params.host()[5] = 0.101;
        cout << ", Sample CPU: " << con_params.host()[5] << "\n";
        //con_params.async_copy_to_device();
        //cout << ", Sample GPU: " << con_params.device()[5] << "\n";
```
The output is:

```
******** Layer 79 ********
con	 (num_filters=32, nr=7, nc=7, stride_y=2, stride_x=2, padding_y=0, padding_x=0) learning_rate_mult=0.1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
con size: 4736
n: 4736, k: 1, nc: 1, nr: 1, Sample CPU: 0.143296
```

I have a few quick questions for you. Thanks again for your great help. I really appreciate it.

1. Why is the tensor size is 4736, not input_channel 3 x num_filters 32 x nr 7 x nr 7 = 4704?
2. I thought k=32, nc=7, nc=7, but the printout gives k=1, nc=1, nr=1. Am I missing anything?
3. If I want to manually a parameter, is it enough to do it with `con_params1.host()[5] = 0.101`? Do I have to update both host and device memory (`con_params1.device()[5] = 0.101`)?
4. I tried to output GPU memory using `con_params.device()[5]` (the last two lines of my code, commented out), but I got segfault. Could you point out where I did wrong?

Sorry about keeping bugging you. Dlib is a great lib. I really like it. Thanks for your help!

**Comment by facug91** [comment]
About number 1, you are forgetting about the bias, which is the same as the number of filters.

**Comment by xiaolin-cheng** [comment]
Thanks Facundo. @facug91 

Any idea if the first 32 elements of a parameter tensor are the bias or the last 32? Thank you!

**Comment by ksachdeva** [comment]
The last 32 elements are that of bias
